### PR TITLE
Allow complex boolean expressions in projection

### LINKED ISSUE #2450: HasValue and != null incorrect translation to sql
I have a linq query that fails in sql because .HasValue or != null gets translated into IS NOT NULL rather than case statements.

This may reflect a misunderstanding on my part, but this seems to be the case.

linq query (note the fee_base, fee_discount, fee_latefee, those values can come from a team, and if not there the age group the team belongs to, and if not there the league the age group belongs to.  This is a coalesce, or isnull scenario except that 0 is allowed so have to test, can't coalesce)

    List<Job_AvailableTeamRegistration_Detail> listJobAvailableTeams_Detail = await (
                from t in db.teams
                join j in db.Jobs on t.jobID equals j.jobID
                join ag in db.agegroups on t.agegroupID equals ag.agegroupID
                join jl in db.Job_Leagues on ag.leagueID equals jl.leagueID
                where
                    t.jobID == cd.JobParams.id
                    && jl.jobID == cd.JobParams.id
                    && t.active == true
                    && t.bAllowSelfRostering == true
                select new Job_AvailableTeamRegistration_Detail()
                {
                    dob_min = t.dobMin ?? ag.dobMin ?? null,
                    dob_max = t.dobMax ?? ag.dobMax ?? null,
                    fee_deposit = t.perRegistrantDeposit ?? 0,
                    fee_base = (t.perRegistrantFee.HasValue && t.perRegistrantFee > 0) ? t.perRegistrantFee : (ag.rosterFee.HasValue && ag.rosterFee > 0) ? ag.rosterFee : 0,
                    fee_discount = (t.discountFee.HasValue && t.discountFee > 0 && DateTime.Now > t.discountFeeStart && DateTime.Now < t.discountFeeEnd) ? t.discountFee : (ag.discountFee.HasValue && ag.discountFee > 0 && DateTime.Now > ag.discountFeeStart && DateTime.Now < ag.discountFeeEnd) ? ag.discountFee : (jl.discountFee.HasValue && jl.discountFee > 0 && DateTime.Now > jl.discountFeeStart && DateTime.Now < jl.discountFeeEnd) ? jl.discountFee : 0,
                    fee_latefee = (t.lateFee != null && t.lateFee > 0 && DateTime.Now > t.lateFeeStart && DateTime.Now < t.lateFeeEnd) ? t.lateFee : (ag.lateFee != null && ag.lateFee > 0 && DateTime.Now > ag.lateFeeStart && DateTime.Now < ag.lateFeeEnd) ? ag.lateFee : (jl.lateFee != null && jl.lateFee > 0 && DateTime.Now > jl.lateFeeStart && DateTime.Now < jl.lateFeeEnd) ? jl.lateFee : 0,
                    gender = t.gender,
                    bAddProcessingFees = j.bAddProcessingFees,
                    fee_processing = 0,
                    grad_year_max = t.grad_year_max ?? ag.grad_year_max ?? null,
                    grad_year_min = t.grad_year_min ?? ag.grad_year_min ?? null,
                    roster_count = (from c in listTeamIDCounts where c.teamID == t.teamID select c.playerCount).SingleOrDefault(),
                    roster_max = t.maxCount,
                    school_grade_max = t.school_grade_max ?? ag.school_grade_max ?? null,
                    school_grade_min = t.school_grade_min ?? ag.school_grade_min ?? null,
                    teamID = t.teamID,
                    teamName = t.teamName
                }
            ).ToListAsync();

SQL SERVER 2014 sql translation:

    exec sp_executesql N'SELECT [t].[dobMin], [ag].[dobMin], NULL, [t].[dobMax], [ag].[dobMax], NULL, [t].[perRegistrantDeposit], [t].[perRegistrantFee] IS NOT NULL, [t].[perRegistrantFee], [ag].[rosterFee] IS NOT NULL, [ag].[rosterFee], [t].[discountFee] IS NOT NULL, [t].[discountFee], @__Now_6 > [t].[discountFeeStart], @__Now_7 < [t].[discountFeeEnd], [ag].[discountFee] IS NOT NULL, [ag].[discountFee], @__Now_9 > [ag].[discountFeeStart], @__Now_10 < [ag].[discountFeeEnd], [jl].[discountFee] IS NOT NULL, [jl].[discountFee], @__Now_12 > [jl].[discountFeeStart], @__Now_13 < [jl].[discountFeeEnd], [t].[lateFee] IS NOT NULL, [t].[lateFee], @__Now_16 > [t].[lateFeeStart], @__Now_17 < [t].[lateFeeEnd], [ag].[lateFee] IS NOT NULL, [ag].[lateFee], @__Now_19 > [ag].[lateFeeStart], @__Now_20 < [ag].[lateFeeEnd], [jl].[lateFee] IS NOT NULL, [jl].[lateFee], @__Now_22 > [jl].[lateFeeStart], @__Now_23 < [jl].[lateFeeEnd], [t].[gender], [j].[bAddProcessingFees], [t].[grad_year_max], [ag].[grad_year_max], NULL, [t].[grad_year_min], [ag].[grad_year_min], NULL, [t].[teamID], [t].[maxCount], [t].[school_grade_max], [ag].[school_grade_max], NULL, [t].[school_grade_min], [ag].[school_grade_min], NULL, [t].[teamName]
    FROM [Leagues].[teams] AS [t]
    INNER JOIN [Jobs].[Jobs] AS [j] ON [t].[jobID] = [j].[jobID]
    INNER JOIN [Leagues].[agegroups] AS [ag] ON [t].[agegroupID] = [ag].[agegroupID]
    INNER JOIN [Jobs].[Job_Leagues] AS [jl] ON [ag].[leagueID] = [jl].[leagueID]
    WHERE ([t].[jobID] = @__cd_JobParams_id_0) AND ([jl].[jobID] = @__cd_JobParams_id_1) AND ([t].[active] = 1) AND ([t].[bAllowSelfRostering] = 1)',N'@__Now_6 datetime,@__Now_7 datetime,@__Now_9 datetime,@__Now_10 datetime,@__Now_12 datetime,@__Now_13 datetime,@__Now_16 datetime,@__Now_17 datetime,@__Now_19 datetime,@__Now_20 datetime,@__Now_22 datetime,@__Now_23 datetime,@__cd_JobParams_id_0 uniqueidentifier,@__cd_JobParams_id_1 uniqueidentifier',@__Now_6='2015-06-23 09:08:52.960',@__Now_7='2015-06-23 09:08:52.960',@__Now_9='2015-06-23 09:08:52.960',@__Now_10='2015-06-23 09:08:52.960',@__Now_12='2015-06-23 09:08:52.960',@__Now_13='2015-06-23 09:08:52.960',@__Now_16='2015-06-23 09:08:52.963',@__Now_17='2015-06-23 09:08:52.963',@__Now_19='2015-06-23 09:08:52.963',@__Now_20='2015-06-23 09:08:52.963',@__Now_22='2015-06-23 09:08:52.963',@__Now_23='2015-06-23 09:08:52.963',@__cd_JobParams_id_0='2ACEA123-D57A-E311-BF81-F04DA202060B',@__cd_JobParams_id_1='2ACEA123-D57A-E311-BF81-F04DA202060B'

### PR DESCRIPTION
**Pull request by smitpatel**
resolves #2450 
Removed all `context.Database.EnsureCreated/Deleted` from `Seed` method. All fixtures call Seed method after dropping/creating database.

### CODE REVIEW

File: test/EntityFramework.MicrosoftSqlServer.FunctionalTests/GearsOfWarQuerySqlServerTest.cs [hunk 1]
```diff
@@ -606,6 +606,60 @@ public override void Select_ternary_operation_with_inverted_boolean()
                 Sql);
         }
 
+        public override void Select_ternary_operation_with_has_value_not_null()
+        {
+            // TODO: Optimize this query
+            base.Select_ternary_operation_with_has_value_not_null();
```
Comment (smitpatel) [inline test/EntityFramework.MicrosoftSqlServer.FunctionalTests/GearsOfWarQuerySqlServerTest.cs]: filed #4267 
Will also write it in comment.

File: src/EntityFramework.MicrosoftSqlServer/Query/Sql/Internal/SqlServerQuerySqlGenerator.cs [hunk 2]
```diff
@@ -132,7 +137,9 @@ protected override Expression VisitBinary(BinaryExpression node)
 
             protected override Expression VisitConditional(ConditionalExpression node)
             {
+                _insideConditionalTest = true;
                 var test = Visit(node.Test);
+                _insideConditionalTest = false;
```
Comment (anpete) [inline src/EntityFramework.MicrosoftSqlServer/Query/Sql/Internal/SqlServerQuerySqlGenerator.cs]: Does it matter if we re-enter here via nested conditionals?

Comment (smitpatel) [inline src/EntityFramework.MicrosoftSqlServer/Query/Sql/Internal/SqlServerQuerySqlGenerator.cs]: Case block inside `When` part (`node.Test`) is not supported. There can be nested conditionals inside `IfTrue` & `IfFalse` but not inside `Test`. So we do not enter VisitConditional while `_insideConditionalTest` is true.
Also added test for nested conditionals.

**Comment by smitpatel** [comment]
Updated

**Comment by anpete** [comment]
:shipit:

**Pull request merged**